javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
I am losing track what version we discuss. This one is incorrect:

```
private volatile Supplier<? extends T> supplier;  // dropping volatile here make it even worse
private T value;

@Override
public T get() {
    if (value == null) { // <--- non-volatile read, this is the data race
        synchronized (this) {
            if (supplier != null) {
                value = supplier.get();
                supplier = null;
            }
        }
    }
    return value;
}
```
You *have* to read supplier (and null-check it) before reading value to get happens-before through supplier write/read.

This one is correct:

```
private volatile Supplier<? extends T> supplier; // should be volatile!
private T value;

@Override
public T get() {
    if (supplier != null) { // <--- volatile read
        synchronized (this) {
            if (supplier != null) {
                value = supplier.get();
                supplier = null;
            }
        }
    }
    return value;
}
```